connect-location/src/main/java/com/ifttt/location/ConnectionRefresher.java

```diff
@@ -65,24 +75,79 @@ public Result doWork() {
         return Result.success();
     }
 
-    public static void schedule(Context context, String connectionId) {
+    static void schedule(Context context, String connectionId) {
         Logger.log("Schedule connection polling");
         WorkManager workManager = WorkManager.getInstance(context);
-        Constraints constraints = new Constraints.Builder().setRequiredNetworkType(NetworkType.CONNECTED).build();
         workManager.enqueueUniquePeriodicWork(WORK_ID_CONNECTION_POLLING,
             ExistingPeriodicWorkPolicy.KEEP,
             new PeriodicWorkRequest.Builder(ConnectionRefresher.class,
                 CONNECTION_REFRESH_POLLING_INTERVAL,
                 TimeUnit.HOURS
-            ).setConstraints(constraints)
-                .setInputData(new Data.Builder().putString(INPUT_DATA_CONNECTION_ID, connectionId).build())
+            ).addTag("connection_id:"
+                + connectionId)
+                .setConstraints(networkConstraint())
+                .setInputData(connectionIdData(connectionId))
                 .build()
         );
     }
 
-    public static void cancel(Context context) {
+    static void executeIfExists(Context context) {
+        Logger.log("Checking periodic connection refresh job");
+
+        WorkManager workManager = WorkManager.getInstance(context);
+        ListenableFuture<List<WorkInfo>> future = workManager.getWorkInfosForUniqueWork(WORK_ID_CONNECTION_POLLING);
+        future.addListener(() -> {
+            try {
+                List<WorkInfo> workInfoList = future.get();
+                // If the refresher job is scheduled, we should only have one WorkInfo.
+                scheduleOneTimeRefreshWork(workManager, workInfoList);
+            } catch (ExecutionException e) {
+                if (BuildConfig.DEBUG) {
+                    e.printStackTrace();
+                }
+            } catch (InterruptedException e) {
+                if (BuildConfig.DEBUG) {
+                    e.printStackTrace();
+                }
+            }
```

Comment:
can we just catch `catch(Exception e)` instead?

Reply:
We want to be explicit here since the method throws these two types of exceptions. We should make sure other types of exceptions can be surfaced during development.